crates/goose/src/agents/base.rs

```diff
@@ -0,0 +1,401 @@
+use async_trait::async_trait;
+use tokio::sync::Mutex;
+
+use super::Agent;
+use crate::providers::base::{Provider, ProviderUsage};
+use crate::register_agent;
+use crate::systems::System;
+
+/// Base implementation of an Agent
+pub struct BaseAgent {
+    systems: Vec<Box<dyn System>>,
+    provider: Box<dyn Provider>,
+    provider_usage: Mutex<Vec<ProviderUsage>>,
+}
+
+impl BaseAgent {
+    pub fn new(provider: Box<dyn Provider>) -> Self {
+        Self {
+            systems: Vec::new(),
+            provider,
+            provider_usage: Mutex::new(Vec::new()),
+        }
+    }
+
+    pub fn add_system(&mut self, system: Box<dyn System>) {
+        self.systems.push(system);
+    }
+}
+
+#[async_trait]
+impl Agent for BaseAgent {
```

Comment:
nit: i found the naming here a bit confusing. it might be cause i was thinking "base" -> "base class" in python. its an implementation of the Agent trait. maybe `AgentImpl` or `AgentImplBase`?

Reply:
What about DefaultAgent instead of BaseAgent? The Impl in the names confuses me a bit, but if that's standard in Rust, happy to do that

Reply:
i like `DefaultAgent`